source/loader/layers/sanitizer/asan_interceptor.cpp

```diff
@@ -0,0 +1,617 @@
+    assert(Res == UR_RESULT_SUCCESS);
+
+    std::vector<char> KernelNameBuf(KernelNameSize);
+    Res = context.urDdiTable.Kernel.pfnGetInfo(
+        Kernel, UR_KERNEL_INFO_FUNCTION_NAME, KernelNameSize,
+        KernelNameBuf.data(), nullptr);
+    assert(Res == UR_RESULT_SUCCESS);
+
+    return std::string(KernelNameBuf.data(), KernelNameSize - 1);
+}
+
+ur_result_t SanitizerInterceptor::allocShadowMemory(
+    ur_context_handle_t Context, std::shared_ptr<DeviceInfo> &DeviceInfo) {
+    if (DeviceInfo->Type == DeviceType::CPU) {
+        context.logger.error("Unsupport device type");
+        return UR_RESULT_ERROR_INVALID_ARGUMENT;
+    } else if (DeviceInfo->Type == DeviceType::GPU_PVC) {
+        /// SHADOW MEMORY MAPPING (PVC, with CPU 47bit)
+        ///   Host/Shared USM : 0x0              ~ 0x0fff_ffff_ffff
+        ///   ?               : 0x1000_0000_0000 ~ 0x1fff_ffff_ffff
+        ///   Device USM      : 0x2000_0000_0000 ~ 0x3fff_ffff_ffff
+        constexpr size_t SHADOW_SIZE = 1ULL << 46;
+
+        // TODO: Protect Bad Zone
+        auto Result = context.urDdiTable.VirtualMem.pfnReserve(
+            Context, nullptr, SHADOW_SIZE, (void **)&DeviceInfo->ShadowOffset);
+        if (Result != UR_RESULT_SUCCESS) {
+            context.logger.error(
+                "Shadow memory is allocated failed on PVC ({})", Result);
+            return Result;
+        }
+
+        DeviceInfo->ShadowOffsetEnd = DeviceInfo->ShadowOffset + SHADOW_SIZE;
+    } else {
+        context.logger.error("Unsupport device type");
+        return UR_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+    context.logger.info("Device ShadowOffset: {} - {}",
+                        (void *)DeviceInfo->ShadowOffset,
+                        (void *)DeviceInfo->ShadowOffsetEnd);
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t SanitizerInterceptor::enqueueMemSetShadow(
+    ur_context_handle_t Context, ur_device_handle_t Device,
+    ur_queue_handle_t Queue, uptr Ptr, uptr Size, u8 Value,
+    ur_event_handle_t DepEvent, ur_event_handle_t *OutEvent) {
+
+    uint32_t NumEventsInWaitList = DepEvent ? 1 : 0;
+    const ur_event_handle_t *EventsWaitList = DepEvent ? &DepEvent : nullptr;
+
+    auto ContextInfo = getContextInfo(Context);
+    auto DeviceInfo = ContextInfo->getDeviceInfo(Device);
+
+    if (DeviceInfo->Type == DeviceType::CPU) {
+        context.logger.error("Unsupport device type");
+        return UR_RESULT_ERROR_INVALID_ARGUMENT;
+    } else if (DeviceInfo->Type == DeviceType::GPU_PVC) {
+        ur_event_handle_t InternalEvent{};
+        ur_event_handle_t *Event = OutEvent ? OutEvent : &InternalEvent;
+
+        uptr ShadowBegin = MemToShadow_PVC(DeviceInfo->ShadowOffset, Ptr);
+        uptr ShadowEnd =
+            MemToShadow_PVC(DeviceInfo->ShadowOffset, Ptr + Size - 1);
+
+        // Maybe in future, we needn't to map physical memory manually
+        const bool IsNeedMapPhysicalMem = true;
```

Comment:
I suggest to remove this variable since at the moment it isn't clear if the `future` case exists.

Reply:
> I suggest to remove this variable since at the moment it isn't clear if the `future` case exists.

I've synced with Michal, system USM will be supported in GPU in the future. At that time, I think we won't need to map physical memory manually.

Reply:
I don't think system USM is a viable solution for shadow memory because it is too slow.

Reply:
I'm not sure. Maybe it will help make Asan more powerful.

Reply:
> I suggest to remove this variable since at the moment it isn't clear if the `future` case exists.

I still suggest removing the code. Even if you insist that system USM can be used for shadow memory, which I don't agree, you can always add it back when IsNeedMapPhysicalMem becomes used in the future.

Reply:
done